scipy/optimize/_bracket.py

```diff
@@ -378,3 +380,277 @@ def customize_result(res, shape):
+        The maximum number of iterations of the algorithm to perform. The number
+        of function evaluations is 3 greater than the number of iterations.
+
+    Returns
+    -------
+    res : _RichResult
+        An instance of `scipy._lib._util._RichResult` with the following
+        attributes. The descriptions are written as though the values will be
+        scalars; however, if `func` returns an array, the outputs will be
+        arrays of the same shape.
+
+        xl, xm, xr : float
+            The left, middle, and right points of the bracket, if the algorithm
+            terminated successfully.
+        fl, fm, fr : float
+            The function value at the left, middle, and right points of the bracket.
+        nfev : int
+            The number of function evaluations required to find the bracket.
+        nit : int
+            The number of iterations of the algorithm that were performed.
+        status : int
+            An integer representing the exit status of the algorithm.
+
+            - ``0`` : The algorithm produced a valid bracket.
+            - ``-1`` : The bracket expanded to the allowable limits. Assuming
+                     unimodality, this implies the endpoint at the limit is a
+                     minimizer.
+            - ``-2`` : The maximum number of iterations was reached.
+            - ``-3`` : A non-finite value was encountered.
+            - ``-4`` : Iteration was terminated by `callback`.
+            - ``1`` : The algorithm is proceeding normally (in `callback` only).
+
+        success : bool
+            ``True`` when the algorithm terminated successfully (status ``0``).
+
+    Notes
+    -----
+    Similar to `scipy.optimize.bracket`, this function seeks to find real
+    points ``xl < xm < xr`` such that ``f(xl) >= f(xm)`` and ``f(xr) >= f(xm)``,
+     where at least one of the inequalities is strict. Unlike `scipy.optimize.bracket`,
+    this function can operate in a vectorized manner on array input, so long as
+    the input arrays are broadcastable with each other. Also unlike
+    `scipy.optimize.bracket`, users may specify minimum and maximum endpoints
+    for the desired bracket.
+
+    Given an initial trio of points `xl = xl0``, ``xm = xm0``, ``xr = xr0``,
+    the algorithm checks if these points already give a valid bracket. If not,
+    a new endpoint, `w` is chosen in the "downhill" direction, `xm` becomes the new
+    opposite endpoint, and either `xl` or `xr` becomes the new middle point,
+    depending on which direction is downhill. The algorithm repeats from here.
+
+    The new endpoint `w` is chosen differently depending on whether or not a
+    boundary `xmin` or `xmax` has been set in the downhill direction. Without
+    loss of generality, suppose the downhill direction is to the right, so that
+    ``f(xl) > f(xm) > f(xr)``. If there is no boundary to the right, then `w`
+    is chosen to be ``xr + factor * (xr - xm)`` where `factor` is controlled by
```

Comment:
I think this means that by default `xr - xm` triples in size each step. I'm not sure if this matters, and I'm not sure what I want, but consider whether this should be `xm + factor * (xr - xm)`? I believe that in `_bracket_root` the total bracket width increases by `factor` each step`. If `(xr - xm) == (xm - xl)`, then I suppose what you have could be interpreted in the same way, but that's not guaranteed, so I'm not sure what is best.

Reply:
Huh. It looks like SciPy's existing scalar minimization bracket finder does what you suggest

https://github.com/scipy/scipy/blob/fb4c65fdf89866728a9762b47e1c75c03e0b2777/scipy/optimize/_optimize.py#L2984

(It uses the `a < c < b` notation, but `c` isn't public.) I think I got mixed up because another reference I was checking uses `a < b < c` notation for what is essentially the same basic algorithm. I'll change it to `xm + factor * (xr - xm)`

Reply:
I'm not sure which update step is actually better, but it seems to work as is.

Reply:
Hah ok let's table this. I don't think it matters much; it's just something that I think is worth discussing in real-time at some point.